File: cdk/lib/spring-boot-eks-stack.ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as eks from 'aws-cdk-lib/aws-eks';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as ecr from 'aws-cdk-lib/aws-ecr';

export class SpringBootEksStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a VPC for our cluster
    const vpc = new ec2.Vpc(this, 'SpringBootEksVpc', {
      maxAzs: 3,
      natGateways: 1, // Use 1 NAT Gateway to save costs
    });

    // Create an ECR repository for our Spring Boot application
    const repository = new ecr.Repository(this, 'SpringBootRepository', {
      repositoryName: 'spring-boot-k8s',
      removalPolicy: cdk.RemovalPolicy.DESTROY, // NOT recommended for production
      imageScanOnPush: true,
    });

    // Create an EKS cluster
    const cluster = new eks.Cluster(this, 'SpringBootCluster', {
      version: eks.KubernetesVersion.V1_27,
      vpc,
      defaultCapacity: 2, // Default capacity of 2 nodes
      defaultCapacityInstance: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.MEDIUM
      ),
      clusterName: 'spring-boot-k8s-cluster',
    });

    // Add Datadog Operator to the cluster
    cluster.addHelmChart('DatadogOperator', {
      chart: 'datadog-operator',
      repository: 'https://helm.datadoghq.com',
      namespace: 'datadog',
      createNamespace: true,
      values: {
        // These values would need to be customized for your Datadog account
        datadog: {
          apiKey: '${DATADOG_API_KEY}', // Replace with your Datadog API key or use AWS Secrets Manager
          appKey: '${DATADOG_APP_KEY}', // Replace with your Datadog APP key or use AWS Secrets Manager
          site: 'datadoghq.com',
          apm: {
            enabled: true,
          },
          logs: {
            enabled: true,
            containerCollectAll: true,
          },
          processAgent: {
            enabled: true,
          },
          dogstatsd: {
            useHostPort: true,
          },
        },
      },
    });

    // Deploy our Spring Boot application to the cluster
    const appLabel = { app: 'spring-boot-k8s' };
    
    const deployment = {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      metadata: {
        name: 'spring-boot-k8s',
        labels: appLabel,
      },
      spec: {
        replicas: 2,
        selector: { matchLabels: appLabel },
        strategy: {
          type: 'RollingUpdate',
          rollingUpdate: {
            maxSurge: 1,
            maxUnavailable: 0,
          },
        },
        template: {
          metadata: {
            labels: appLabel,
            annotations: {
              'ad.datadoghq.com/java.logs': '[{"source":"java","service":"spring-boot-k8s"}]',
            },
          },
          spec: {
            containers: [{
              name: 'spring-boot-k8s',
              image: `${repository.repositoryUri}:latest`,
              ports: [{ containerPort: 8080 }],
              resources: {
                requests: {
                  memory: '512Mi',
                  cpu: '200m',
                },
                limits: {
                  memory: '1Gi',
                  cpu: '500m',
                },
              },
              readinessProbe: {
                httpGet: {
                  path: '/health',
                  port: 8080,
                },
                initialDelaySeconds: 30,
                periodSeconds: 10,
              },
              livenessProbe: {
                httpGet: {
                  path: '/health',
                  port: 8080,
                },
                initialDelaySeconds: 60,
                periodSeconds: 20,
              },
              env: [
                {
                  name: 'JAVA_OPTS',
                  value: '-XX:+UseContainerSupport -XX:MaxRAMPercentage=75.0',
                },
                {
                  name: 'DD_AGENT_HOST',
                  valueFrom: {
                    fieldRef: {
                      fieldPath: 'status.hostIP',
                    },
                  },
                },
                {
                  name: 'DD_ENV',
                  value: 'production',
                },
                {
                  name: 'DD_SERVICE',
                  value: 'spring-boot-k8s',
                },
                {
                  name: 'DD_VERSION',
                  value: '1.0.0',
                },
                {
                  name: 'DD_LOGS_INJECTION',
                  value: 'true',
                },
                {
                  name: 'DD_TRACE_SAMPLE_RATE',
                  value: '1',
                },
                {
                  name: 'DD_PROFILING_ENABLED',
                  value: 'true',
                },
              ],
              volumeMounts: [{
                name: 'config-volume',
                mountPath: '/app/config',
              }],
            }],
            volumes: [{
              name: 'config-volume',
              configMap: {
                name: 'spring-boot-k8s-config',
              },
            }],
          },
        },
      },
    };

    const service = {
      apiVersion: 'v1',
      kind: 'Service',
      metadata: {
        name: 'spring-boot-k8s',
        labels: appLabel,
      },
      spec: {
        type: 'LoadBalancer', // Using LoadBalancer to expose the service externally
        ports: [{
          port: 80,
          targetPort: 8080,
          protocol: 'TCP',
          name: 'http',
        }],
        selector: appLabel,
      },
    };

    const configMap = {
      apiVersion: 'v1',
      kind: 'ConfigMap',
      metadata: {
        name: 'spring-boot-k8s-config',
      },
      data: {
        'application.properties': `
spring.application.name=spring-boot-k8s
server.port=8080

# Actuator configuration
management.endpoints.web.exposure.include=health,info,metrics
management.endpoint.health.show-details=always

# Datadog configuration
dd.service.name=spring-boot-k8s
dd.env=production
dd.version=1.0.0
        `,
      },
    };

    // Apply Kubernetes manifests to the cluster
    cluster.addManifest('SpringBootApp', deployment, service, configMap);

    // Output the ECR repository URI
    new cdk.CfnOutput(this, 'RepositoryUri', {
      value: repository.repositoryUri,
      description: 'ECR Repository URI for the Spring Boot application',
    });

    // Output the EKS cluster name
    new cdk.CfnOutput(this, 'ClusterName', {
      value: cluster.clusterName,
      description: 'EKS Cluster Name',
    });

    // Output the kubectl role ARN
    new cdk.CfnOutput(this, 'KubectlRoleArn', {
      value: cluster.kubectlRole?.roleArn || '',
      description: 'IAM Role ARN for kubectl access',
    });
  }
}

File: cdk-java/src/com/twinspires/cdk/SpringBootEksStack.java
package com.twinspires.cdk;

import software.amazon.awscdk.Duration;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.ec2.Vpc;
import software.amazon.awscdk.services.ec2.SubnetConfiguration;
import software.amazon.awscdk.services.ec2.SubnetType;
import software.amazon.awscdk.services.ecr.Repository;
import software.amazon.awscdk.services.eks.Cluster;
import software.amazon.awscdk.services.eks.KubernetesManifest;
import software.amazon.awscdk.services.eks.KubernetesVersion;
import software.amazon.awscdk.services.eks.NodegroupOptions;
import software.amazon.awscdk.services.eks.CapacityType;
import software.amazon.awscdk.services.eks.EndpointAccess;
import software.amazon.awscdk.services.iam.AccountRootPrincipal;
import software.constructs.Construct;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SpringBootEksStack extends Stack {
    public SpringBootEksStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        // Create a VPC for our cluster
        Vpc vpc = Vpc.Builder.create(this, "SpringBootVpc")
                .maxAzs(2)
                .subnetConfiguration(Arrays.asList(
                        SubnetConfiguration.builder()
                                .name("public")
                                .subnetType(SubnetType.PUBLIC)
                                .cidrMask(24)
                                .build(),
                        SubnetConfiguration.builder()
                                .name("private")
                                .subnetType(SubnetType.PRIVATE_WITH_EGRESS)
                                .cidrMask(24)
                                .build()
                ))
                .build();

        // Create an ECR repository for our container images
        Repository ecrRepository = Repository.Builder.create(this, "SpringBootRepository")
                .repositoryName("spring-boot-k8s")
                .removalPolicy(software.amazon.awscdk.RemovalPolicy.DESTROY)
                .build();

        // Create an EKS cluster
        Cluster cluster = Cluster.Builder.create(this, "SpringBootCluster")
                .version(KubernetesVersion.V1_27)
                .vpc(vpc)
                .defaultCapacity(0) // We will define the nodegroup separately
                .endpointAccess(EndpointAccess.PUBLIC_AND_PRIVATE)
                .clusterName("spring-boot-eks")
                .build();

        // Add IAM user access to the cluster
        cluster.awsAuth().addMastersRole(new AccountRootPrincipal());

        // Add a managed nodegroup
        cluster.addNodegroupCapacity("SpringBootNodeGroup", NodegroupOptions.builder()
                .instanceTypes(Arrays.asList(
                        new software.amazon.awscdk.services.ec2.InstanceType("t3.medium")
                ))
                .minSize(2)
                .maxSize(4)
                .desiredSize(2)
                .capacityType(CapacityType.ON_DEMAND)
                .build());

        // Define Kubernetes namespace
        Map<String, Object> namespaceManifest = new HashMap<>();
        namespaceManifest.put("apiVersion", "v1");
        namespaceManifest.put("kind", "Namespace");
        Map<String, Object> namespaceMetadata = new HashMap<>();
        namespaceMetadata.put("name", "spring-boot");
        namespaceManifest.put("metadata", namespaceMetadata);

        KubernetesManifest namespace = KubernetesManifest.Builder.create(this, "SpringBootNamespace")
                .cluster(cluster)
                .manifest(Collections.singletonList(namespaceManifest))
                .build();

        // Define Kubernetes deployment
        Map<String, Object> deploymentManifest = createDeploymentManifest();
        KubernetesManifest deployment = KubernetesManifest.Builder.create(this, "SpringBootDeployment")
                .cluster(cluster)
                .manifest(Collections.singletonList(deploymentManifest))
                .build();
        deployment.node().addDependency(namespace);

        // Define Kubernetes service
        Map<String, Object> serviceManifest = createServiceManifest();
        KubernetesManifest service = KubernetesManifest.Builder.create(this, "SpringBootService")
                .cluster(cluster)
                .manifest(Collections.singletonList(serviceManifest))
                .build();
        service.node().addDependency(deployment);
    }

    private Map<String, Object> createDeploymentManifest() {
        Map<String, Object> manifest = new HashMap<>();
        manifest.put("apiVersion", "apps/v1");
        manifest.put("kind", "Deployment");

        Map<String, Object> metadata = new HashMap<>();
        metadata.put("name", "spring-boot-k8s");
        metadata.put("namespace", "spring-boot");
        manifest.put("metadata", metadata);

        Map<String, Object> spec = new HashMap<>();
        spec.put("replicas", 2);

        Map<String, Object> selector = new HashMap<>();
        Map<String, Object> matchLabels = new HashMap<>();
        matchLabels.put("app", "spring-boot-k8s");
        selector.put("matchLabels", matchLabels);
        spec.put("selector", selector);

        Map<String, Object> template = new HashMap<>();
        Map<String, Object> templateMetadata = new HashMap<>();
        Map<String, Object> labels = new HashMap<>();
        labels.put("app", "spring-boot-k8s");
        templateMetadata.put("labels", labels);
        template.put("metadata", templateMetadata);

        Map<String, Object> templateSpec = new HashMap<>();
        List<Map<String, Object>> containers = Arrays.asList(createContainerSpec());
        templateSpec.put("containers", containers);
        template.put("spec", templateSpec);
        spec.put("template", template);
        manifest.put("spec", spec);

        return manifest;
    }

    private Map<String, Object> createContainerSpec() {
        Map<String, Object> container = new HashMap<>();
        container.put("name", "spring-boot-k8s");
        container.put("image", "${ECR_REPOSITORY_URI}:latest");

        List<Map<String, Object>> ports = Arrays.asList(
                Collections.singletonMap("containerPort", 8080)
        );
        container.put("ports", ports);

        Map<String, Object> resources = new HashMap<>();
        Map<String, Object> limits = new HashMap<>();
        limits.put("cpu", "1000m");
        limits.put("memory", "1024Mi");
        Map<String, Object> requests = new HashMap<>();
        requests.put("cpu", "500m");
        requests.put("memory", "512Mi");
        resources.put("limits", limits);
        resources.put("requests", requests);
        container.put("resources", resources);

        List<Map<String, Object>> env = Arrays.asList(
                createEnvVar("SPRING_PROFILES_ACTIVE", "prod"),
                createEnvVar("DD_APM_ENABLED", "true"),
                createEnvVar("DD_LOGS_ENABLED", "true"),
                createEnvVar("DD_PROCESS_AGENT_ENABLED", "true")
        );
        container.put("env", env);

        Map<String, Object> livenessProbe = new HashMap<>();
        Map<String, Object> httpGet = new HashMap<>();
        httpGet.put("path", "/actuator/health");
        httpGet.put("port", 8080);
        livenessProbe.put("httpGet", httpGet);
        livenessProbe.put("initialDelaySeconds", 60);
        livenessProbe.put("periodSeconds", 10);
        container.put("livenessProbe", livenessProbe);

        Map<String, Object> readinessProbe = new HashMap<>();
        Map<String, Object> readinessHttpGet = new HashMap<>();
        readinessHttpGet.put("path", "/actuator/health");
        readinessHttpGet.put("port", 8080);
        readinessProbe.put("httpGet", readinessHttpGet);
        readinessProbe.put("initialDelaySeconds", 30);
        readinessProbe.put("periodSeconds", 5);
        container.put("readinessProbe", readinessProbe);

        return container;
    }

    private Map<String, Object> createEnvVar(String name, String value) {
        Map<String, Object> env = new HashMap<>();
        env.put("name", name);
        env.put("value", value);
        return env;
    }

    private Map<String, Object> createServiceManifest() {
        Map<String, Object> manifest = new HashMap<>();
        manifest.put("apiVersion", "v1");
        manifest.put("kind", "Service");

        Map<String, Object> metadata = new HashMap<>();
        metadata.put("name", "spring-boot-k8s");
        metadata.put("namespace", "spring-boot");
        manifest.put("metadata", metadata);

        Map<String, Object> spec = new HashMap<>();
        spec.put("type", "LoadBalancer");

        Map<String, Object> selector = new HashMap<>();
        selector.put("app", "spring-boot-k8s");
        spec.put("selector", selector);

        List<Map<String, Object>> ports = Arrays.asList(
                createServicePort("http", 80, 8080)
        );
        spec.put("ports", ports);
        manifest.put("spec", spec);

        return manifest;
    }

    private Map<String, Object> createServicePort(String name, int port, int targetPort) {
        Map<String, Object> servicePort = new HashMap<>();
        servicePort.put("name", name);
        servicePort.put("port", port);
        servicePort.put("targetPort", targetPort);
        return servicePort;
    }
}
